### DISCUSSION TITLE
How to filter Enum hybrid property? (expression generates 'WHERE false')

### DISCUSSION THREAD

**Discussion by RamonGiovane** [sqlalchemy/sqlalchemy#8518] [Usage Questions] [answered]
It's actually a complex use of case with joined inheritance and polymorphism.

I have a column that is an `Enum` and it's value is assigned by a hybrid setter.

Here's how it's defined, sorta:
```python
Base = declarative_base()

class ReportType(str, enum.Enum):
   scientific_report = 'scientific_report'
   adf = 'ADF'

  
class ReviewArea(str, enum.Enum):
  foo_committee = 'foo_committee'
  bar_department =  'bar_department'

class BaseReport(Base):
    __tablename__ = 'base_report'
  

    id = Column(Integer, primary_key=True)

    description = Column(String)
    
    
    report_type = Column(Enum(ReportType), nullable=False)

    __mapper_args__ = {
        'polymorphic_on': report_type
    }

    # Here!
    _review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        self._review_area = review_area

class ScientificReport(BaseReport):
    __tablename__ = 'scientific_report'

    id = Column(Integer, ForeignKey('base_report.id'), primary_key=True)

    something_scientific = Column(String)
    
   _review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        BaseReport._review_area = review_area
        self._review_area = review_area

    __mapper_args__ = {
        'polymorphic_identity': ReviewType.scientifc_report,
    }
```

#### Little explanation:
I needed to set it this way, I will not provide all the code here, because it would be boring to read, but it's basically a copy, both  child and parent classes have a `review_area` column.
- I repeated the column in the subclasses because I needed to set specific constraints on the child `review_area`s (DB level).
- I kept the column in the base class because I need to query data from a certain `review_area`
-  I know that redundancy is something to avoid, but I think it would have better performance (it would be too much ` LEFT OUTER JOINS` to get a summary of reports with some `ReviewArea`), the queries would be less painful to write, and the data safety will not be compromised.

Anyways, insertions with the above code work fine, with a RDBMS I could assert that it's populating both parent and child classes with their corresponding review_area:
```python
session.add(ScientificReport(review_area=ReviewArea.foo_committee)
session.commit()
```
No errors nor warnings.

#### But:
```python
select(BaseReport).where(BaseReport.review_area == ReviewArea.foo_committee)`
```

does not work. When executing this, the generated SQL statement replaces this `where` with:

`WHERE false`

I tried to add some hybrid `expression`. Tried something like:
```python
@review_area.expression
def review_area(cls):
  return cls._review_area
```
It did not work. the same thing happened!
It's a shoot in the dark, I admit. I did not found a use case with enums like this and I have no clue in what to do.

#### EDIT:
If I query by the child class, it works great:
```python
select(ScientificReport).where(ScientificReport.review_area == ReviewArea.foo_committee)
```
But the point is to filter by the base class, without loading or even knowing who are the children classes.

**Comment by CaselIT** [comment]
Hi,

I've not tried the example to see what's happening, but by just looking at it, it seems that the hybrid property is not needed.
Does it work if you remove it?

**Reply by RamonGiovane** [reply]
I edited the topic, I was forgetting some points, but I think it makes more sense now, take a look, please.

**Comment by zzzeek** [comment, marked as answer]
the hybrid looks fine, and i can't reproduce your problem.  modify the test script below to get it to do what you're seeing:


```py

import enum

from sqlalchemy import Column
from sqlalchemy import Enum, select
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import declarative_base

Base = declarative_base()

class ReviewArea(str, enum.Enum):
  foo_committee = 'foo_committee'
  bar_department =  'bar_department'

class BaseReport(Base):
    __tablename__ = 'base_report'


    id = Column(Integer, primary_key=True)

    description = Column(String)


    __mapper_args__ = {
        'polymorphic_on': description,
        'with_polymorphic': '*',
    }

    # Here!
    _review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        self._review_area = review_area

class ScientificReport(BaseReport):
    __tablename__ = 'scientific_report'

    id = Column(Integer, ForeignKey('base_report.id'), primary_key=True)

    something_scientific = Column(String)

    _review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        self._review_area = review_area

    __mapper_args__ = {
        'polymorphic_identity': "ADF",
    }


print(select(ScientificReport).where(ScientificReport.review_area == ReviewArea.foo_committee))
```

output

```sql
SELECT scientific_report.review_area, base_report.review_area AS review_area_1, scientific_report.id, base_report.id AS id_1, base_report.description, scientific_report.something_scientific 
FROM base_report JOIN scientific_report ON base_report.id = scientific_report.id 
WHERE scientific_report.review_area = :review_area_2
```

**Reply by RamonGiovane** [reply]
Oh, I'm sorry, I forgot two things, I'll update the topic.
First thing is, the child setter also sets the base property, in order to make the value available on both parent and child tables.
```python
    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        BaseReport._review_area = review_area
        self._review_area = review_area

    __mapper_args__ = {
        'polymorphic_identity': "ADF",
    }
```
The second thing is, I typed the wrong query :facepalm: 
What I'm trying to achieve is:
```python
select(BaseReport).where(BaseReport.review_area == ReviewArea.foo_committee)
```
I wanted to filter by the base class, without loading or even knowing who are the children.
The above evaluates to `WHERE = false`

```python
select(ScientificReport).where(ScientificReport.review_area == ReviewArea.foo_committee)
```
This, in the other hand, works good, by the way :+1: 

Another thing I'm doing wrong is using `'with_polymorphic': '*'` since I don't want to load everything all the time :smile:

**Reply by CaselIT** [reply]
> First thing is, the child setter also sets the base property, in order to make the value available on both parent and child tables.

The error is here. This `BaseReport._review_area = review_area` is overriding the value of the model `BaseReport._review_area` from the sqlalchemy descriptior that can be used in the query setting it to the enum value.
From the setter of the child class you don't have access to the parent instance of BaseReport. You can make it work if the child instance is inside a session already and you query the session to obtain the parent instance from it

Since `BaseReport.review_area` is an enum value when you are doing the query, python will evaluate to True / False the `==`

**Reply by RamonGiovane** [reply]
I see... makes sense!

Just out of curiosity, I've put a `breakpoint()`  in the child settter.
Before the override, the stored value of `BaseReport._review_area` was `<sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x7fd432942e30>`, and then it becomes `<ReviewArea.scientific_report: 'scientific_report'>` lol

The crazy thing is, in DB layer, the column gets populated. 
But I feel like I'm doing some nasty hacking I shouldn't.

If there isn't  any other way I might abandon this idea and stick with validators and have a `review_area` in the base table only :confused:

**Reply by RamonGiovane** [reply]
I think the discussion got a little bit out of track, since the problem is not really with the enum expression. As you guys mentioned, I don't even need it. 
But just for the record, I managed to get  what I wanted working!
Thanks for the support guys, you're the best!

```python
Base = declarative_base()

class ReportType(str, enum.Enum):
   scientific_report = 'scientific_report'
   adf = 'ADF'

  
class ReviewArea(str, enum.Enum):
  foo_committee = 'foo_committee'
  bar_department =  'bar_department'

class BaseReport(Base):
    __tablename__ = 'base_report'
  

    id = Column(Integer, primary_key=True)

    description = Column(String)
    
    
    report_type = Column(Enum(ReportType), nullable=False)

    __mapper_args__ = {
        'polymorphic_on': report_type
    }

    # Using different name, because it's from a different table!
    _base_review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._base_review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        self._base_review_area = review_area

class ScientificReport(BaseReport):
    __tablename__ = 'scientific_report'

    id = Column(Integer, ForeignKey('base_report.id'), primary_key=True)

    something_scientific = Column(String)
    
   _review_area = Column('review_area', Enum(ReviewArea), index=True, nullable=False)

    @hybrid_property
    def review_area(self) -> ReviewArea:
        return self._review_area

    @review_area.setter
    def review_area(self, review_area: ReviewArea):
        # The intention here is not to override, but have the same value
        # populated both in parent and child tables
        self._base_review_area = review_area
        self._review_area = review_area

    __mapper_args__ = {
        'polymorphic_identity': ReviewType.scientifc_report,
    }

    # Now I can do my specific constraint business without increasing query complexity
    __table_args__ = (
        Index(
            "ix_unique_some_scientific_review_area",
            "something_scientific",
            "review_area",
            unique=True,
        ),
    )
```

Now, both of the below statements work fine. The generated SQL queries are correct.
```python
select(BaseReport).where(BaseReport.review_area == ReviewArea.foo_committee)

select(ScientificReport).where(ScientificReport.review_area == ReviewArea.foo_committee)
```

**Comment by CaselIT** [comment]
Just to add that this may be due to the use of `is` instead of `==`. That cannot be used with queries.

So look at cases where the code is using `.where(ScientificReport.review_area is ReviewArea.foo_committee)` instead of `.where(ScientificReport.review_area == ReviewArea.foo_committee)`

**Reply by RamonGiovane** [reply]
I'm using `==`, but this is good to know. Thank you.
